app/controllers/registration/business/ReceiveGoodsNIPController.scala

```diff
@@ -0,0 +1,73 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.registration.business
+
+import config.{AuthClientConnector, BaseControllerComponents, FrontendAppConfig}
+import connectors.KeystoreConnector
+import controllers.BaseController
+import forms.ReceiveGoodsNIPForm
+import models.api.NETP
+import play.api.mvc.{Action, AnyContent}
+import services._
+import views.html.NorthernIrelandReceiveGoods
+
+import javax.inject.Inject
+import scala.concurrent.ExecutionContext
+
+class NorthernIrelandReceiveGoodsController @Inject()(val keystoreConnector: KeystoreConnector,
+                                                      val authConnector: AuthClientConnector,
+                                                      val applicantDetailsService: ApplicantDetailsService,
+                                                      val returnsService: ReturnsService,
+                                                      val tradingService: TradingDetailsService,
+                                                      val vatRegistrationService: VatRegistrationService,
+                                                      view: NorthernIrelandReceiveGoods)
+                                                     (implicit appConfig: FrontendAppConfig,
+                                                      val executionContext: ExecutionContext,
+                                                      baseControllerComponents: BaseControllerComponents)
+  extends BaseController with SessionProfile {
+
+  def show: Action[AnyContent] = isAuthenticatedWithProfile() {
+    implicit request =>
+      implicit profile =>
+        for {
+          northernIrelandReceiveGoodsView <- returnsService.getReturns.map(returns => returns.northernIrelandProtocol.map(nip => nip.goodsFromEU))
+          form = ReceiveGoodsNIPForm.fillWithPrePop(northernIrelandReceiveGoodsView)
+        } yield Ok(view(form))
+  }
+
+  def submit: Action[AnyContent] = isAuthenticatedWithProfile() {
+    implicit request =>
+      implicit profile =>
+        ReceiveGoodsNIPForm.form.bindFromRequest.fold(
+          errors =>
+            for {
+              _ <- returnsService.getReturns.map(returns => returns.northernIrelandProtocol)
+            } yield BadRequest(view(errors)),
+          success => {
+            val (receiveGoods, amount) = success
+            returnsService.saveNorthernIrelandReceiveGoods(receiveGoods, amount) flatMap { _ =>
```

Comment:
To avoid creating tons of single use functions in the returns service could you use the .submitReturns one, like in the SendEUGoodsController for example